### ISSUE TITLE
Exclusive Registration issues

### ISSUE THREAD

**Issue by jason-fox** [ScorpioBroker/ScorpioBroker#449]
Testing with Scorpio 4.1.6

Create a Registration without a type:

```json
  {
       
        "endpoint": "http://iot-agent:4041",
...
        "mode": "exclusive",
        "operations": "updateOps"
    }
```

Error message states: 

```json
"detail": {
        "message": "A registration needs type which is CSourceRegistration"
    }
```

This is incorrect -  **Table 5.2.9-1: CSourceRegistration data type definition** states:

>  type String It shall be equal to `"ContextSourceRegistration" `

Create an **exclusive** registration with multiple commands:

```console
curl -L 'http://localhost:1026/ngsi-ld/v1/csourceRegistrations/' \
-H 'Content-Type: application/json' \
-H 'Link: <http://context/ngsi-context.jsonld>; rel="http://www.w3.org/ns/json-ld#context"; type="application/ld+json"' \
-H 'NGSILD-Tenant: openiot' \
-d '   {
        "type": "ContextSourceRegistration",
        "contextSourceInfo": {
            "key": "jsonldContext",
            "value": "http://context/ngsi-context.jsonld"
        },
        "endpoint": "http://iot-agent:4041",
        "information": [
            {
                "entities": [
                    {
                        "id": "urn:ngsi-ld:Device:filling111",
                        "type": "FillingLevelSensor"
                    }
                ],
                "propertyNames": [
                    "add",
                    "remove",
                    "refill"
                ]
            }
        ],
        "mode": "exclusive",
        "operations": "updateOps"
    }'
```

On retrieval:

```console
curl -L 'http://localhost:1026/ngsi-ld/v1/csourceRegistrations/?type=FillingLevelSensor' \
-H 'Accept: application/ld+json' \
-H 'Link: <http://context/ngsi-context.jsonld>; rel="http://www.w3.org/ns/json-ld#context"; type="application/ld+json"' \
-H 'NGSILD-Tenant: openiot'
```

The response contains **duplicated** entries for the registration 

<summary>
   Open for details:
   <details>

```json
  {
        "id": "urn:ngsi-ld:csourceregistration:-1996389586",
        "type": "ContextSourceRegistration",
        "contextSourceInfo": {
            "key": "jsonldContext",
            "value": "http://context/ngsi-context.jsonld"
        },
        "endpoint": "http://iot-agent:4041",
        "information": [
            {
                "entities": [
                    {
                        "id": "urn:ngsi-ld:Device:filling111",
                        "type": "FillingLevelSensor"
                    }
                ],
                "propertyNames": [
                    "add",
                    "remove",
                    "refill"
                ]
            }
        ],
        "mode": "exclusive",
        "operations": "updateOps",
        "@context": [
            "http://context/ngsi-context.jsonld",
            "https://uri.etsi.org/ngsi-ld/v1/ngsi-ld-core-context-v1.7.jsonld"
        ]
    },
    {
        "id": "urn:ngsi-ld:csourceregistration:-1996389586",
        "type": "ContextSourceRegistration",
        "contextSourceInfo": {
            "key": "jsonldContext",
            "value": "http://context/ngsi-context.jsonld"
        },
        "endpoint": "http://iot-agent:4041",
        "information": [
            {
                "entities": [
                    {
                        "id": "urn:ngsi-ld:Device:filling111",
                        "type": "FillingLevelSensor"
                    }
                ],
                "propertyNames": [
                    "add",
                    "remove",
                    "refill"
                ]
            }
        ],
        "mode": "exclusive",
        "operations": "updateOps",
        "@context": [
            "http://context/ngsi-context.jsonld",
            "https://uri.etsi.org/ngsi-ld/v1/ngsi-ld-core-context-v1.7.jsonld"
        ]
    },
    {
        "id": "urn:ngsi-ld:csourceregistration:-1996389586",
        "type": "ContextSourceRegistration",
        "contextSourceInfo": {
            "key": "jsonldContext",
            "value": "http://context/ngsi-context.jsonld"
        },
        "endpoint": "http://iot-agent:4041",
        "information": [
            {
                "entities": [
                    {
                        "id": "urn:ngsi-ld:Device:filling111",
                        "type": "FillingLevelSensor"
                    }
                ],
                "propertyNames": [
                    "add",
                    "remove",
                    "refill"
                ]
            }
        ],
        "mode": "exclusive",
        "operations": "updateOps",
        "@context": [
            "http://context/ngsi-context.jsonld",
            "https://uri.etsi.org/ngsi-ld/v1/ngsi-ld-core-context-v1.7.jsonld"
        ]
    }
]
```
</details>
<summary>

Actuating a command for a registered exclusive device:

```console
curl -L -X PATCH 'http://localhost:1026/ngsi-ld/v1/entities/urn:ngsi-ld:Device:filling001/attrs/remove' \
-H 'Link: <https://fiware.github.io/tutorials.Step-by-Step/tutorials-context.jsonld>; rel="http://www.w3.org/ns/json-ld#context"; type="application/ld+json"' \
-H 'Content-Type: application/json' \
-H 'NGSILD-Tenant: openiot' \
-d '{ 
	"type": "Property", 
	"value": " "
} '
```

Returns an error as shown:

```json
{
    "type": "https://uri.etsi.org/ngsi-ld/errors/InternalError",
    "title": "Internal error",
    "detail": {
        "message": "Cannot invoke \"io.smallrye.mutiny.tuples.Tuple2.getItem1()\" because \"splitted\" is null"
    },
    "errorCode": 500
}
```

Actuating a device on the default tenant, when a device has been registered on a named tenant:

```console
curl -L -X PATCH 'http://localhost:1026/ngsi-ld/v1/entities/urn:ngsi-ld:Device:filling001/attrs/remove' \
-H 'Link: <https://fiware.github.io/tutorials.Step-by-Step/tutorials-context.jsonld>; rel="http://www.w3.org/ns/json-ld#context"; type="application/ld+json"' \
-H 'Content-Type: application/json' \
-d '{ 
	"type": "Property", 
	"value": " "
} '
```

Results in an error:

```json
{
    "type": "https://uri.etsi.org/ngsi-ld/errors/InternalError",
    "title": "Internal error",
    "detail": {
        "message": null
    },
    "errorCode": 500
}
```

If I have an exclusive registration, I can check that the registration exists, e.g: for `urn:ngsi-ld:Device:water001`

```json
"information": [
            {
                "entities": [
                    {
                        "id": "urn:ngsi-ld:Device:water001",
                        "type": "Water"
                    }
                ],
                "propertyNames": [
                    "on",
                    "off"
                ]
            }
        ],
        "mode": "exclusive",
        "operations": "updateOps",
```

However,  an attempt to actuate that entity results in an internal error:

```console
curl -L -X PATCH 'http://localhost:1026/ngsi-ld/v1/entities/urn:ngsi-ld:Device:water001/attrs/off' \
-H 'Link: <https://fiware.github.io/tutorials.Step-by-Step/tutorials-context.jsonld>; rel="http://www.w3.org/ns/json-ld#context"; type="application/ld+json"' \
-H 'Content-Type: application/json' \
-H 'NGSILD-Tenant: openiot' \
-d '{ 
	"type": "Property", 
	"value": " "
} '
```

```text
2023-10-05 15:37:44,531 ERROR [eu.nec.ngs.com.too.HttpUtils] (vert.x-eventloop-thread-73) Exception :: : java.util.NoSuchElementException
```

I assume this is because no entity data is found within the broker itself, however that is a valid situation for a **redirect** or **exclusive** registration.

**Comment by ScorpioBroker** [comment]
Hi,

mostly addressed but we need to find a way to add this kind of setup in our ci

**Comment by jason-fox** [comment]
I'm trying to create a new tutorial based on the various actuation types, so I'm using this as a basis:

https://github.com/FIWARE/tutorials.Context-Providers/tree/NGSI-LD

Once it is ready it should hold all 4 registration types, but I want to run using a variety of brokers, federated brokers, partial brokers, IoT Agents etc.

I actually discovered this issue #449 trying to replicate and close #427 by running the following:

https://github.com/FIWARE/tutorials.Subscriptions/tree/NGSI-LD

I just updated `.env` to hold Scorpio and IoT Agent to latest and ran `./services scorpio` and the device provisioning is not working as expected - there is a 409 somewhere because of duplicated provisioning.  I then ran a manual registration to replicate the issue.
